core/services/keystore/models.go

```diff
@@ -423,3 +424,68 @@ func (rawKeys rawKeyRing) keys() (*keyRing, error) {
 func adulteratedPassword(password string) string {
 	return "master-password-" + password
 }
+
+type ResourceMutex struct {
+	mu          sync.Mutex
+	activeCount map[ServiceType]int // Tracks active users per service type
+}
+type ServiceType int
+
+const (
+	TXMv1 ServiceType = iota
+	TXMv2
+)
+
+// TryLock attempts to lock the resource for the specified service type.
+// It returns an error if the resource is locked by a different service type.
+func (rm *ResourceMutex) TryLock(serviceType ServiceType) error {
+	rm.mu.Lock()
+	defer rm.mu.Unlock()
+
+	// Check if other service types are using the resource
+	for otherServiceType, count := range rm.activeCount {
+		if otherServiceType != serviceType && count > 0 {
+			return errors.New("resource is locked by another service type")
+		}
+	}
+
+	// Increment active count for the current service type
+	rm.activeCount[serviceType]++
+	return nil
+}
+
+// Unlock releases the lock for the service type
+func (rm *ResourceMutex) Unlock(serviceType ServiceType) error {
+	rm.mu.Lock()
+	defer rm.mu.Unlock()
+
+	// Check if the service type has an active lock
+	if rm.activeCount[serviceType] == 0 {
+		return errors.New("no active lock for this service type")
+	}
+
+	// Decrement active count for the service type
+	rm.activeCount[serviceType]--
+	if rm.activeCount[serviceType] == 0 {
+		delete(rm.activeCount, serviceType)
+	}
+	return nil
+}
+
+// IsLocked checks if the resource is locked by any service or a specific service type.
```

Comment:
As written currently, it does not appear to check for 'any service'. Does it need to?
```suggestion
// IsLocked checks if the resource is locked by a specific service type.
```

Reply:
Fixed in [`759f3e1` (#15969)](https://github.com/smartcontractkit/chainlink/pull/15969/commits/759f3e19731a3dfb4a3b74f81686313c508caabd)